lib/src/main/java/org/openbase/bco/dal/lib/action/ActionDescriptionProcessor.java

```diff
@@ -219,11 +219,36 @@ public static ActionReference generateActionReference(final ActionDescriptionOrB
      * @return the initial initiator of an action as described above.
      */
     public static ActionInitiator getInitialInitiator(final ActionDescriptionOrBuilder actionDescription) {
-        if (actionDescription.getActionCauseList().isEmpty()) {
-            return actionDescription.getActionInitiator();
-        } else {
-            return actionDescription.getActionCause(actionDescription.getActionCauseCount() - 1).getActionInitiator();
+        return getInitialActionReference(actionDescription).getActionInitiator();
+    }
+
+    /**
+     * Get an action reference to the last action of the cause chain.
+     * If the action does not have a cause, an action reference for the provided action description is returned.
+     *
+     * @param actionDescription the action description from which the original action reference is resolved.
+     *
+     * @return the initial reference of an action.
+     */
+    public static ActionReference getInitialActionReference(final ActionDescriptionOrBuilder actionDescription) {
+
+        final long now = System.currentTimeMillis();
+
+        // return the first action that is still valid.
+        for (int i = actionDescription.getActionCauseList().size() - 1; i >= 0; i--) {
+            try {
+            final ActionReference actionCause = actionDescription.getActionCause(i);
+            final long pastTime = (now - TimestampProcessor.getTimestamp(actionCause, TimeUnit.MILLISECONDS));
+                if(pastTime <= getExecutionTimePeriod(actionCause, TimeUnit.MICROSECONDS)) {
+                    return actionCause;
```

Comment:
* maybe add to the documentation that actions no longer valid by their execution time period are returned?
* wouldn't it be better to have a dedicated method to check if an action is still in its execution time if there is none yet?